### PR TITLE
tds/odbc: unified execute + statement-wise result navigation API (msodbcsql parity)

### PR DESCRIPTION
**Pull request by David-Engel**
## Summary
Replaces the split execute/result-navigation surface in `mssql-tds` with a single unified API, and makes statement-wise result navigation (msodbcsql parity) the default across every consumer.

This started (superseding #141) as *additive* statement-wise `SQLMoreResults` navigation for `mssql-odbc`. In review we decided that, since these APIs have no external consumers to preserve, a clean breaking redesign is better than carrying two parallel navigation paths forever. This PR now delivers that unified API and migrates all consumers plus the full test suite.

## New unified API (`mssql-tds`)
- `execute(sql, options) -> StatementResult`, where `options: impl Into<ExecuteOptions>` so `()` selects defaults.
- `StatementResult`: `Rows` / `NoRows { rows_affected }` / `End`.
- `advance() -> StatementResult` — drain the current statement and move to the next statement boundary.
- `advance_to_rows() -> bool` — collapse forward to the next row-returning statement (preserves the old collapsing behavior where a caller wants it).
- `on_rows()` and `has_open_batch()` accessors.
- `ExecuteOptions { timeout, cancel, column_encryption }` (with `Default` + builder methods) replaces the trailing `timeout, cancel` (and encryption-setting) parameters on every `execute_*` method.

**msodbcsql rule** (verified by probing the reference driver): a statement is a navigable result iff it returns rows, carries a row count (DONE `COUNT` flag), or produced messages (INFO). Pure no-op statements (e.g. a bare `CREATE TABLE`) collapse. Predicate: `has_count || saw_message`.

## Breaking changes
Removed in favor of the unified surface:
- `execute_multi_statement`, `move_to_next_statement`
- the `ResultSetClient` trait (`get_current_resultset` / `move_to_next`)
- `execute_sp_executesql_with_encryption_setting` / `_core`
- the `expose_norow` bool on the result-boundary loop — statement-wise navigation is now unconditional.

Every `execute_*` method now takes `impl Into<ExecuteOptions>` instead of positional `timeout` / `cancel` args.

## Consumers migrated
`mssql-odbc`, `mssql-js`, `mssql-py-core`, `mssql-tds-cli`, and `mssql-tds-bench`, plus the full `mssql-tds` unit + integration test suite (~40 files).

## Production fix
`get_next_row_into`'s error arm now clears `has_open_batch` / result-set state after draining, so `close_query` / `advance` no longer hang after a mid-row server error.

## New test scaffolding: `mssql-tds/src/test_client_support.rs`
This PR adds a small `test-util`-gated module that builds a `TdsClient` on top of a mock transport which replays a *scripted* sequence of TDS tokens (`COLMETADATA`, `DONE`/`DONEPROC` with specific flags, `INFO`) instead of talking to a live server. It is not shipped API and is excluded from coverage reporting.

**Why it's needed.** The unified statement-wise navigation pushes real logic into the seam between the TDS layer and the ODBC layer — no-row results, message-only statements, end-of-batch, and the defensive error guards. Several of those paths are impossible or flaky to reach through a real server:
- Internal error guards can't be produced by a real server — e.g. an ODBC statement left cursor-open while its connection has no active client, or a transport that dies *mid-drain*.
- Some token shapes are awkward to coax out of SQL Server — e.g. a zero-column `COLMETADATA`, or an exact `DONE` flag combination such as `MORE|COUNT`.
- These are unit tests: they run in the Build stage with **no SQL host**, and they assert on driver-internal state (STMT/DBC handle state, `SQLSTATE`s, cursor open/closed) that the black-box ODBC C API does not expose.

**Benefits.**
- Deterministic, hermetic coverage of the ODBC result-navigation state machine and its error branches — no server, no network flakiness.
- Fast feedback in the Build stage, complementing the live integration tests and the e2e `--compare-with-msodbcsql` parity run, which cover the happy paths end-to-end.

**Why it lives in `mssql-tds` behind a feature.** Constructing a positioned `TdsClient` requires crate-internal transport traits (`TdsTransport`, `TdsTokenStreamReader`, …) and `TdsClient::new`. Exposing these helpers through the `test-util` feature — opted into by `mssql-odbc`'s dev-dependency — is the idiomatic Rust pattern; a separate testkit crate would instead force widening many internals to `pub`.

## Tests & parity
- Whole workspace builds; `clippy` + `fmt` clean.
- `mssql-tds` lib + integration suites, `mssql-odbc` lib, and the mock / cursor / rpc / tvp / reset-connection / bulk-copy suites are green (local SQL Server 2022; VECTOR and a timing-flaky table-lock case aside, which are environment/flake, not regressions).
- **e2e `run_e2e.sh --compare-with-msodbcsql`: 14 parity / 0 bugs.** The comparison caught one real regression — prepared DML left a 0-column cursor open (SQLSTATE `24000` on reuse) — fixed by draining non-`Rows` prepared results to idle in the `SQLExecute` path.

### CODE REVIEW

File: mssql-tds/src/connection/tds_client.rs [hunk 1]
```diff
@@ -1764,17 +1782,38 @@ impl TdsClient {
...
+    /// next COLMETADATA — consecutive no-row statements collapse into the
+    /// following row-returning result set.
+    ///
+    /// With `true` (ODBC statement-wise navigation, matching msodbcsql), each
+    /// statement's DONE token is its own result boundary: a no-row statement is
+    /// returned as [`ResultBoundaryKind::NoRows`] instead of being skipped, so
+    /// every statement in a batch is individually navigable. A DONE reached in
+    /// this method (without a COLMETADATA earlier in the same call) always
+    /// belongs to a no-row statement, because a row-returning statement's DONE
+    /// is consumed while its rows are read/drained.
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs]: Fixed in 71786a3c: the doc now states no-row statements surface only when they carry a row count (DONE COUNT flag) or produced a message, and that pure no-op statements (e.g. bare CREATE TABLE) are still collapsed.

File: mssql-tds/src/connection/tds_client.rs [hunk 2]
```diff
@@ -1883,20 +1943,118 @@ impl TdsClient {
...
+
+    /// Executes a SQL batch and positions on its **first statement** using
+    /// statement-wise navigation, returning that statement's [`StatementResult`].
+    ///
+    /// Unlike [`execute()`](Self::execute) — which skips leading no-row
+    /// statements to position on the first row-returning result set — this
+    /// exposes every statement (including PRINT / RAISERROR / DDL / DML) as its
+    /// own navigable result, matching msodbcsql's `SQLExecDirect` +
+    /// `SQLMoreResults` semantics. Advance through the remaining statements with
+    /// [`move_to_next_statement()`](Self::move_to_next_statement).
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs]: Fixed in 71786a3c: doc updated to 'first navigable statement' and documents the count-or-message predicate, noting a leading bare CREATE TABLE is collapsed so the first result may belong to a later statement.

File: mssql-tds/src/connection/tds_client.rs [hunk 3]
```diff
@@ -3137,6 +3295,43 @@ impl ResultSetClient for TdsClient {
...
 
+/// The outcome of advancing to the next statement boundary during
+/// statement-wise result navigation
+/// ([`execute_multi_statement`](TdsClient::execute_multi_statement) /
+/// [`move_to_next_statement`](TdsClient::move_to_next_statement)).
+///
+/// Unlike the result-set navigation used by [`execute`](TdsClient::execute) /
+/// [`move_to_next`](TdsClient::move_to_next) — which collapses statements that
+/// return no rows — this exposes every statement in a batch individually,
+/// matching msodbcsql's `SQLExecDirect` + `SQLMoreResults` behavior.
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs]: Fixed in 71786a3c: the enum and NoRows variant docs now state the count-or-message rule and that pure no-op statements never surface as NoRows.

File: mssql-tds/src/connection/tds_client.rs [hunk 4]
```diff
@@ -1883,20 +1943,118 @@ impl TdsClient {
...
+        }
+        if self.maybe_has_unread_rows() {
+            self.drain_rows().await?;
+        }
+        // Draining the current result set may have consumed the batch's final
+        // DONE token (has_more=false), which closes the batch. If so there is
+        // nothing left on the wire to advance to; reading again would block
+        // forever waiting for a token that never arrives.
+        if !self.execution_context.has_open_batch() {
+            return Ok(StatementResult::End);
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs:2060]: Added in 71786a3c: move_to_next_statement_end_after_draining_final_rowset exercises a final row set (COLMETADATA + terminal DONE), advances without reading rows, and asserts End under a 5s timeout. The mock transport's receive_row_into now replays queued tokens so the drain path is covered; without the post-drain re-check the test fails (errors instead of returning End).

File: mssql-py-core/src/cursor.rs [hunk 5]
```diff
@@ -61,13 +61,22 @@ impl PyCoreCursor {
...
+                        },
+                    )
+                    .await
+                    .map_err(|e| {
+                        error!("execute: Failed to execute query: {}", e);
+                        pyo3::exceptions::PyRuntimeError::new_err(format!(
+                            "Query execution failed: {}",
+                            e
+                        ))
+                    })?;
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-py-core/src/cursor.rs]: Fixed in 32dd3635: `Cursor.execute` now advances to the first row-returning result set (`advance_to_rows`) when the unified execute stops on a no-row boundary, so `CREATE; INSERT; SELECT` (and any leading DDL/DML) exposes the SELECT to `fetchone`/`fetchall` again — restoring the pre-statement-wise collapsing behavior.

File: mssql-js/src/connection.rs [hunk 6]
```diff
@@ -110,9 +110,7 @@ impl Connection {
             .collect();
 
         let rpc_params = rpc_params?;
-        let result = client
-            .execute_sp_executesql(query, rpc_params, None, None)
-            .await;
+        let result = client.execute_sp_executesql(query, rpc_params, ()).await;
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-js/src/connection.rs]: Fixed in 32dd3635: `executeWithParams` (line 113) and `executeProc` (line 155) — plus the plain `execute` — now collapse to the first row set after executing, matching `query_raw`. So `UPDATE ...; SELECT ...` no longer loses the SELECT when `createResultFast`/`fetchChunk` runs.

File: mssql-tds/src/test_client_support.rs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        true
+    }
+}
+
+/// Builds a [`TdsClient`] whose transport replays `tokens`. Combine with the
+/// public statement-wise navigation API
+/// ([`TdsClient::execute_multi_statement`](crate::connection::tds_client::TdsClient::execute_multi_statement)
+/// / [`move_to_next_statement`](crate::connection::tds_client::TdsClient::move_to_next_statement))
+/// to position the client on a scripted result before handing it to a
+/// consumer under test.
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/test_client_support.rs:178]: Fixed in 32dd3635: the `test_client_support` doc comment now links to the unified `TdsClient::execute`/`advance` API. Also repointed the remaining stale links to the removed `execute_multi_statement`/`move_to_next_statement`/`ResultSetClient` (in `lib.rs` quick-start doctest — now compiles — plus `cursor_ops.rs`, `query.rs`, and `tds_client.rs`).

**Comment by Theekshna** [comment]
Looks good. Would be good to make sure we're covering most edge cases etc in out e2e/unit tests

**Review by Theekshna — APPROVED** [review]

File: mssql-tds/tests/common/mod.rs [hunk 8]
```diff
@@ -151,38 +151,38 @@ pub async fn validate_results(
...
-                        print!("Row {actual_rows:?}: ");
-                        for cell in row {
-                            print!("{cell:?},");
-                        }
-                        println!();
-                        actual_rows += 1;
+    let mut has_rows = client.on_rows();
+    if !has_rows {
+        has_rows = client.advance_to_rows().await?;
+    }
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/tests/common/mod.rs]: Fixed in 445edcdc: `validate_results` is now statement-wise — it takes the `StatementResult` returned by `execute`, iterates with `advance()`, validates `NoRows.rows_affected` against `Update(n)` and row counts against `Result(n)`, and asserts the batch is fully consumed. Pure no-op statements (DDL/SET/ROLLBACK) collapse and are expressed as `Update(0)` (consume no boundary). Verified against query_results/transaction/reset_connection/timeout_and_cancel (39/39).

File: mssql-tds/src/connection/tds_client.rs [hunk 9]
```diff
@@ -601,33 +621,23 @@ impl TdsClient {
         };
 
         self.begin_command();
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs:653]: Fixed in 445edcdc: `begin_command()` now clears `return_values` (not just `info_messages`), so a plain batch executed after a fully-navigated prior RPC no longer reports stale `get_return_values()`. This covers this batch prologue and `execute_sp_executesql`, and the four previously-explicit per-path clears were removed as redundant.

File: mssql-tds/src/connection/tds_client.rs [hunk 10]
```diff
@@ -641,59 +651,22 @@ impl TdsClient {
...
-    ) -> TdsResult<()> {
-        self.current_command_ce_setting = encryption_setting;
+        options: impl Into<ExecuteOptions<'a>>,
+    ) -> TdsResult<StatementResult> {
+        let ExecuteOptions {
+            timeout: timeout_sec,
+            cancel: cancel_handle,
+            column_encryption,
+        } = options.into();
+        self.current_command_ce_setting = column_encryption;
```
[the remark this answers is not in the document]
Comment (David-Engel) [reply mssql-tds/src/connection/tds_client.rs:699]: Fixed in 445edcdc: handled centrally — `begin_command()` now clears `return_values`, so `execute_sp_executesql` no longer exposes stale RPC outputs from a prior command.

**Comment by David-Engel** [comment]
@saurabh500 - This is a breaking API change. But I think it cleans things up nicely and is worth it. Let me know what you think.

**Comment by saurabh500** [comment]
LGTM

**Review by saurabh500 — APPROVED** [review]
